### ISSUE TITLE
RuboCop pulls in .rubocop.yml files from vendor directories

### ISSUE THREAD

**Issue by tas50** [rubocop/rubocop#9832]
## Expected behavior

Configs should not be loaded from vendor directories, which may be created automatically by CI environments like GitHub Actions.

## Actual behavior

Anything in a vendor directory is processed, which causes failures when those directories are cached by CI systems like GitHub Actions.

In our case as we have a folder called kitchen-tests and inside that is a vendor dir since kitchen-tests has it's own Gemfile/Gemfile.lock. When RuboCop runs at the root of the project is ends up attempting to load a config from the vendor dir here which tries to require rubocop-discourse.

```
Unable to find gem rubocop-discourse; is the gem installed? Gem::MissingSpecError
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_loader_resolver.rb:263:in `rescue in gem_config_path'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_loader_resolver.rb:253:in `gem_config_path'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_loader_resolver.rb:69:in `block (2 levels) in resolve_inheritance_from_gems'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_loader_resolver.rb:67:in `reverse_each'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_loader_resolver.rb:67:in `block in resolve_inheritance_from_gems'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_loader_resolver.rb:61:in `each_pair'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_loader_resolver.rb:61:in `resolve_inheritance_from_gems'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_loader.rb:50:in `load_file'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_loader.rb:103:in `configuration_from_file'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_store.rb:68:in `for_dir'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/config_store.rb:58:in `for'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:161:in `ruby_interpreters'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:153:in `ruby_executable?'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:169:in `ruby_file?'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:73:in `to_inspect?'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:64:in `block in target_files_in_dir'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:64:in `select'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:64:in `target_files_in_dir'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:38:in `block in find'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:36:in `each'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/target_finder.rb:36:in `find'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/runner.rb:77:in `find_target_files'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/runner.rb:42:in `run'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/cli/command/execute_runner.rb:26:in `block in execute_runner'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/cli/command/execute_runner.rb:52:in `with_redirect'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/cli/command/execute_runner.rb:25:in `execute_runner'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/cli/command/execute_runner.rb:17:in `run'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/cli/command.rb:11:in `run'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/cli/environment.rb:18:in `run'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/cli.rb:65:in `run_command'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/cli.rb:72:in `execute_runners'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/lib/rubocop/cli.rb:41:in `run'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/exe/rubocop:12:in `block in <top (required)>'
/usr/local/Cellar/ruby@2.7/2.7.3/lib/ruby/2.7.0/benchmark.rb:308:in `realtime'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-1.15.0/exe/rubocop:12:in `<top (required)>'
/usr/local/lib/ruby/gems/2.7.0/bin/rubocop:23:in `load'
/usr/local/lib/ruby/gems/2.7.0/bin/rubocop:23:in `<main>'
```

That is being loaded from here

```
cat kitchen-tests/vendor/bundle/ruby/2.7.0/gems/mini_mime-1.1.0/.rubocop.yml
inherit_gem:
  rubocop-discourse: default.yml
inherit_mode:
  merge:
    - Exclude
```

## Steps to reproduce the problem

Vendor a gem that includes either deprecated configs or requires a plugin and then run RuboCop.

## RuboCop version

```
1.15.0 (using Parser 3.0.1.1, rubocop-ast 1.6.0, running on ruby 2.7.3 x86_64-darwin20)
```

**Comment by dvandersluis** [comment]
Have you tried:

```yaml
AllCops:
  Exclude:
  - vendor/**/*.rb
```

Doing this by default for RuboCop would probably be a breaking change even though people probably shouldn't be running on code in `vendor/`.

**Comment by tas50** [comment]
@dvandersluis I'll give that config a try. It seems like RuboCop should skip configs in a vendor dir even if the vendor dir itself is scanned though. I'm struggling to think of a positive to finding configs here and it seems like a number of projects are accidentally shipping their rubocop configs due to loose gemspecs.

**Comment by tas50** [comment]
That config change doesn't seem to fix it based on this CI run: https://buildkite.com/chef-oss/chef-chefstyle-master-verify/builds/248#d62595f1-42d8-4a86-b0c9-4abc72604d42

**Comment by erikw** [comment]
> Have you tried:
> 
> ```yaml
> AllCops:
>   Exclude:
>   - vendor/**/*.rb # or vendor/bundle/**/*
> ```
> 
> Doing this by default for RuboCop would probably be a breaking change even though people probably shouldn't be running on code in `vendor/`.

I had the same issue as @tas50 and it took me a long time to nail it down.

It turns out that when building with multiple Gemfiles in Travis, e.g. `.travis.yml` snippet:

```yml
gemfile:
  - gemfiles/jekyll_3.7.x.gemfile
  - gemfiles/jekyll_4.x.x.gemfile
cache: bundler
```

there will be a lot of rubocop `configuration.yml` found in `gemfiles/vendor/bundle/*`. When I ran rubocop like this in travis:
```yml
script: bundle exec rubocop --debug
```
I could see from the debug output and I could see that rubocop was picking up files like:

```
configuration from /home/travis/build/$MYPROJ/gemfiles/vendor/bundle/ruby/3.0.0/gems/rubocop-rake-0.6.0/config/default.yml
Default configuration from /home/travis/build/$MYPROJ/gemfiles/vendor/bundle/ruby/3.0.0/gems/rubocop-1.18.2/config/default.yml
configuration from /home/travis/build/$MYPROJ/gemfiles/vendor/bundle/ruby/3.0.0/gems/rubocop-rspec-2.4.0/config/default.yml
```

The problem with this is that these files seems to depend on some Gems that are (naturally) not set up in my Gemfiles. In my case with the above included `default.yml` included, my build would always fail with
```console
$ bundle exec rake
[...] 
Unable to find gem panolint; is the gem installed? Gem::MissingSpecError
```


Why it took quite some time for me to figure out was the fact that it works locally always, and that the `cache: bundler` must be set for the problem to be triggered in Travis, building with normal `Gemfile` in project root does not trigger this on Travis. So I went down the route and configure the build matrix manually an setting the travis envvar `CACHE_NAME`.

I'm glad I found this issue @tas50 and your soution @dvandersluis. Just a few minutes after finding this issue I added to my `.rubocop.yml`:

```yml
AllCops:
  Exclude:
  - gemfiles/vendor/bundle/**/*
```

and the issue I've spent hours and hours on was resolved. Thanks :)

Could it make sense to add some FAQ/common problems to this project to warn about the combination `rubocop + travis + multiple Gemfiles + travis-bundle-cache`?

**Comment by mphsi** [comment]
This issue may also appear when working with GitHub Actions and using the `ruby/setup-ruby@v1` action, which installs and caches the gems automatically.

So, excluding the vendor directory in the `.rubocop.yml` did the trick. In my case I did:

```yml
AllCops:
  Exclude:
    - vendor/bundle/**/*
```

**Comment by dswij** [comment]
Default config seems to [exclude vendor](https://github.com/rubocop/rubocop/blob/9f24166d04922361a84d2e1e95ef4d15f81443e0/config/default.yml#L66). 

At first glance, this seems to be a problem when defining an `Exclude` config under `AllCops`. The new `Exclude` override the vendor pattern defined in the default config. 

I fixed my problem with `inherit_mode` set to `merge` for `Exclude` in  `.rubocop.yml` (because we inherited from a shared config and would not want to override it), but this does not seem to be a pleasant solution. Another solution is to explicitly set all exclude patterns in `.rubocop.yml`

**Comment by jonas054** [comment]
It looks like RuboCop behaved as expected, or at least as specified in its documentation. Closing.

**Comment by dswij** [comment]
Although I agree that it behaved as specified, I think the behavior can be improved somehow, or at least this can be documented better.

**Comment by akostadinov** [comment]
For anybody looking here, I had same issue with github actions. As in the @dswij's comment, this fixed the issue:
```yaml
inherit_mode:
  merge:
    - Exclude
```

**Comment by jpowell** [comment]
For posterity... I was trying to exclude `db/schema.rb` from all cops as to keep the rails generated version and not fail 
CI everytime we made an update to the DB... that's what led me down the path of finding this issue, as we were getting the `Unable to find gem panolint` error... 

The solution mentioned above by @akostadinov and @dswij was what worked for us
```yml
inherit_mode:
  merge:
    - Exclude
```
